File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.cpp
#include <string>
#include <vector>
#include<map>
using namespace std;

vector<string> solution(vector<string> players, vector<string> callings) {
    vector<string> answer;
    map<string,int> ranks;
    for(int i=0;i<players.size();i++){
        ranks.insert({players[i],i});
    }
    for(auto it:callings){
        int rank = ranks[it];
        
        swap(players[rank],players[rank-1]);
        ranks[it]--;
        ranks[players[rank]]++;
    }
    return players;
}

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.java
import java.io.*;
import java.util.*;
class Solution {
    public String[] solution(String[] players, String[] callings) {
        String[] answer = new String[players.length];
        Map<String,Integer> playerSeq = new HashMap<>();
        Map<Integer,String> seq = new HashMap<>();
        for(int i=0;i<players.length;i++){
            playerSeq.put(players[i],i);
            seq.put(i,players[i]);
        }
        for(String it:callings){
            int temp = playerSeq.get(it)-1;
            playerSeq.put(it,temp);
            String prev = seq.get(temp);
            playerSeq.put(prev,temp+1);
            seq.put(temp+1,prev);
            seq.put(temp,it);
        }
        seq.forEach((key,value) -> {
            answer[key] = value; 
        });
        return answer;
    }
}